### PR TITLE
Add functionality for stacking spectrum observations and grouping

### PR DESCRIPTION
**Pull request by JouvinLea**
Starting to implement the grouping but this PR is just to discuss with @joleroi so there is nothing to review yet

### CODE REVIEW

**Comment by JouvinLea** [comment]
@joleroi 
Hi, I am having an error in one of your test in the test_spectral_extraction for the test: test_spectrum_extraction_from_configfile.
```
gammapy/spectrum/tests/test_spectrum_extraction.py::test_spectrum_extraction PASSED
gammapy/spectrum/tests/test_spectrum_extraction.py::test_spectrum_extraction_from_configfile FAILED
gammapy/spectrum/tests/test_spectrum_extraction.py::test_spectrum_extraction_grouping_from_an_observation_list PASSED

=================================== FAILURES ===================================
___________________ test_spectrum_extraction_from_configfile ___________________

tmpdir = local('/private/var/folders/33/gcrqtty96178yy041w7k8y300001_t/T/pytest-jouvin/pytest-4/test_spectrum_extraction_from_0')

    @requires_dependency('yaml')
    @requires_dependency('scipy')
    @requires_data('gammapy-extra')
    def test_spectrum_extraction_from_configfile(tmpdir):
        configfile = gammapy_extra.filename(
            'test_datasets/spectrum/spectrum_analysis_example.yaml')
        config = read_yaml(configfile)
>       config['extraction']['results']['outdir'] = str(tmpdir)
E       KeyError: u'results'

gammapy/spectrum/tests/test_spectrum_extraction.py:89: KeyError
===================== 1 failed, 2 passed in 10.68 seconds ======================
```

I don't see this test in the master branch of gammapy anymore so maybe you changed something?

File: gammapy/spectrum/spectrum_extraction.py [hunk 1]
```diff
@@ -18,6 +16,7 @@
 from ..image import ExclusionMask
 from ..region import SkyCircleRegion, find_reflected_regions
 from ..utils.energy import EnergyBounds, Energy
+from ..irf import EffectiveAreaTable, EnergyDispersion
```
Comment (JouvinLea) [inline gammapy/spectrum/spectrum_extraction.py]: The import error comes from here:
```
running test
running build
running build_py
copying gammapy/spectrum/spectrum_extraction.py -> build/lib.macosx-10.9-x86_64-2.7/gammapy/spectrum
running build_ext
skipping 'gammapy/detect/_test_statistics_cython.c' Cython extension (up-to-date)
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "_pytest.main", line 80, in wrap_session
INTERNALERROR>     config.do_configure()
INTERNALERROR>   File "_pytest.config", line 621, in do_configure
INTERNALERROR>     self.hook.pytest_configure(config=self)
INTERNALERROR>   File "_pytest.core", line 521, in __call__
INTERNALERROR>     return self._docall(self.methods, kwargs)
INTERNALERROR>   File "_pytest.core", line 528, in _docall
INTERNALERROR>     firstresult=self.firstresult).execute()
INTERNALERROR>   File "_pytest.core", line 394, in execute
INTERNALERROR>     res = method(*args)
INTERNALERROR>   File "gammapy/conftest.py", line 40, in pytest_configure
INTERNALERROR>     from .utils.testing import has_data
INTERNALERROR>   File "gammapy/utils/testing.py", line 9, in <module>
INTERNALERROR>     from ..data import DataManager
INTERNALERROR>   File "gammapy/data/__init__.py", line 16, in <module>
INTERNALERROR>     from .spectral_cube import *
INTERNALERROR>   File "gammapy/data/spectral_cube.py", line 18, in <module>
INTERNALERROR>     from ..spectrum import LogEnergyAxis, powerlaw
INTERNALERROR>   File "gammapy/spectrum/__init__.py", line 19, in <module>
INTERNALERROR>     from .spectrum_extraction import *
INTERNALERROR>   File "gammapy/spectrum/spectrum_extraction.py", line 19, in <module>
INTERNALERROR>     from ..irf import EffectiveAreaTable, EnergyDispersion
INTERNALERROR>   File "gammapy/irf/__init__.py", line 9, in <module>
INTERNALERROR>     from .exposure import *
INTERNALERROR>   File "gammapy/irf/exposure.py", line 4, in <module>
INTERNALERROR>     from ..data import SpectralCube
INTERNALERROR> ImportError: cannot import name SpectralCube
```

Comment (cdeil) [inline gammapy/spectrum/spectrum_extraction.py]: The problem is the import here:
https://github.com/gammapy/gammapy/blob/master/gammapy/irf/exposure.py#L4

As a temp solution, you could move this into the exposure_cube function and things should work.

I'll think about a long-term proper solution and make a PR.

File: gammapy/spectrum/spectrum_extraction.py [hunk 2]
```diff
@@ -329,50 +333,75 @@ def from_datastore(cls, obs_id, store, on_region, bkg_method, ebounds,
...
+       Averaged exposure ratio between ON and OFF regions, arf and rmf
 
-        :math:`\\alpha_{\\mathrm{tot}}` for all observations is calculated as
+       :math:`\\alpha_{\\mathrm{tot}}`  for all observations is calculated as
 
-        .. math:: \\alpha_{\\mathrm{tot}} = \\frac{\\sum_{i}\\alpha_i \\cdot N_i}{\\sum_{i} N_i}
+       .. math:: \\alpha_{\\mathrm{tot}} = \\frac{\\sum_{i}\\alpha_i \\cdot N_i}{\\sum_{i} N_i}
 
-        where :math:`N_i` is the number of OFF counts for observation :math:`i`
+       .. math:: \\arf_{\\mathrm{tot}} = \\frac{\\sum_{i}\\arf_i \\cdot \\livetime_i}{\\sum_{i} \\livetime_i}
```
Comment (JouvinLea) [inline gammapy/spectrum/spectrum_extraction.py]: @joleroi 
doesn't work on the sphynx docu... There is a small syntax error no.

**Comment by joleroi** [comment]
@JouvinLea 

I think the easiest solution is if you rebase your branch on the master branch. Since you mostly added a new file this should pass through without big issues.

I will tell you the basics steps I do to make a rebase

* Commit all your changes so that ``git status`` is clean
* Make a backup :wink: 
```
git branch backup
```
* Get newest master
```
git checkout master
git pull
```
* Go back to your branch an rebase master
```
git checkout < your branch>
git rebase master
```
If your're lucky this will pass through, but maybe you will get merge conflicts, that means git does not know how to merge two versions of a file together (changes in master, and your changes).
It will tell you which files is affects and if you open this file you will see the differences marked (with ``>>>>>>>>>>>>>>>>>`` and ``<<<<<<<<<<<<<<<<<<<<<<``).
Edit the file in order to produce a version you think is correct and add it with 
``
git add <file
``
Now you can
``git rebase --continue``
If you're lost at some point do
```
git rebase --abort
```

What happens in the merge process is basically simple. git rewinds all your local changes to the time when you checked out the master branch. Then it adds all changes thath have been done to master by other on your trunkated branch. Then it applies your changes to the newest master version. It's confusing admittedly, but you will get used to it :smile:

If you want to know everything this link can be helpful
https://www.atlassian.com/git/tutorials/merging-vs-rebasing/conceptual-overview

**Comment by JouvinLea** [comment]
@joleroi 
Yes I did that but i have conflict on the file test_spectrum_extraction.py and I don't know how to solve it since after the rebase I don't see this test that I added:
test_spectrum_extraction_grouping_from_an_observation_list()
??

**Comment by joleroi** [comment]
Did you commit your changes before rebasing? If yes, I could imagine that he merge conflict happens for a commit that lies (temporary)  before the commit that includes your new test, and so the test will be added when you've resolved the merge conflict and the commit include the new test is added to master.

But this is just a guess, @adonath @cdeil ideas?

**Comment by cdeil** [comment]
> But this is just a guess, @adonath @cdeil ideas?

No.

@joleroi - Can you fetch this branch now so that you have a backup?

@JouvinLea - If you can't figure out the rebase, Johannes or I can do it and then you continue in a new branch after this is merged in master.

**Comment by joleroi** [comment]
I did the rebase. At one point a didn't know which version to keep so I left both and added a comment (in test_spectrum_extaction.py I think)

you can fetch the branch like this

```
git remote add joleroi https://github.com/joleroi/gammapy.git
git fetch joleroi
git branch -a (see all my branches)
git checkout -b <branch name> remotes/joleroi/grouping_rebased
```

As far as I understand github you should be able to keep this PR by pushing to JouvinLea:SpectralGroupingClass ( you might need the ``-f`` option``)

**Comment by JouvinLea** [comment]
@joleroi 
Ok I have it on a new branch but before using push -f I want to be sure that all my test works after the rebase and this is not the case.
I fixed some stuff since for example the EffectiveAreatable init function changes... I get this erros at the end of the function grouping_from_an_observation_list in SpectrumObservation for the energy_range...
Apparently you changed the keyword?

**Comment by joleroi** [comment]
> @joleroi
Ok I have it on a new branch but before using push -f I want to be sure that all my test works after the rebase and this is not the case.
I fixed some stuff since for example the EffectiveAreatable init function changes... I get this erros at the end of the function grouping_from_an_observation_list in SpectrumObservation for the energy_range...
Apparently you changed the keyword?

You can also push to a new branch (if you want to keep the old one) and open a new PR. This way I or @cdeil  can also have a look at the test failures and help you fix them. I don't know about the EffetiveAreaTable but in general there are a lot of changes in Gammapy at the moment. Once you have your stuff in master (and under test) it will be easy to spot of some change breaks your stuff (and ideally the person who introduces the changes, fixes it).
So I would say we should get something (even if it's preliminary) into master as soon as possible, to avoid having to manually rewrite it all the time.

**Comment by cdeil** [comment]
Superseded by #471.

**Pull request closed without merging**